pytest_catchlog.py

```diff
@@ -162,59 +190,96 @@ def record_tuples(self):
 
             (logger_name, log_level, message)
         """
-        return [(r.name, r.levelno, r.getMessage()) for r in self.records()]
-
-    def set_level(self, level, logger=None):
-        """Sets the level for capturing of logs.
-
-        By default, the level is set on the handler used to capture
-        logs. Specify a logger name to instead set the level of any
-        logger.
-        """
+        return [(r.name, r.levelno, r.getMessage()) for r in self.buffer]
 
-        obj = logger and logging.getLogger(logger) or self.handler
-        obj.setLevel(level)
+    @property
+    def buffer(self):
+        return self._handler.buffer
 
-    def at_level(self, level, logger=None):
-        """Context manager that sets the level for capturing of logs.
+    @property
+    def lines(self):
+        return [r.getMessage() for r in self.buffer]
 
-        By default, the level is set on the handler used to capture
-        logs. Specify a logger name to instead set the level of any
-        logger.
+    def __init__(self, handler):
+        # Do not call LineMatcher.__init__(),
+        # it attempts to assign to self.lines, which is a property.
+        self._handler = handler
+
+    @classmethod
+    @contextmanager
+    def recording(cls, **kwargs):
+        with catching_logs(RecordingHandler(), **kwargs) as handler:
+            yield cls(handler)
+
+    def at_level(self, level):
+        """Context manager that sets the level for recording of logs.
+
+        This only affects logs collected by this funcarg. To control
+        captured logs that are reported for failed tests use the
+        ``logging_at_level()`` context manager.
```

Comment:
That's mainly because there're separate handler instances used for reporting and inspection now. And naively changing a level of a logger through a `at_level` call on some recorder would affect all attached handlers, i.e. other recorders. This looks wrong to me, despite there used to be the same behavior before.

Honestly, I don't know what to do about that. It is possible to use a separate recorder for a certain `foo.bar` logger on some level, and even adjust the level in the middle of recording:
```
with caplog.recording(logging.INFO, logger='foo.bar') as rec1, \
     caplog.recording(logging.INFO, logger='some.other.module') as rec2:

    # Capture INFO and above.
    ...

    with rec1.at_level(logging.WARNING):
        # Capture only WARNING and below.
        # This doesn't affect a level of rec2 though.
        ...
```

But there's no way to change a level on, say, `foo.bar.baz` logger, except for `caplog.logging_at_level()`, which changes the level of the logger itself, affecting other recorders, including log reporting on failure.

I'll think about that. May be we could use log Filters for that purpose. I'd like to preserve backward compatibility after all.